compiler/optimizer/VPHandlers.cpp

```diff
@@ -9744,7 +9744,52 @@ static TR::Node *constrainIfcmpeqne(OMR::ValuePropagation *vp, TR::Node *node, b
        static const char* disableProfiled2Overridden = feGetEnv ("TR_DisableProfiled2Overridden");
        if (!disableProfiled2Overridden && callNode && vp->lastTimeThrough())
           {
-
+          bool isFixedClass = false;
+
+          // If the guard has a method test, then we need to get objectClass
+          // from the child of the lhs node (a load of a vft symbol).
+          // We also need to ensure that the method of the original call node is
+          // not overridden in the hierarchy of the receiver's class.
+          static const char* disableMethodTest2Hierarchy = feGetEnv ("TR_DisableMethodTest2Hierarchy");
+          bool isMethodTest = vp->comp()->getSymRefTab()->isVtableEntrySymbolRef(node->getFirstChild()->getSymbolReference());
+          TR_OpaqueClassBlock* inlinedMethodClass = NULL;
+          if (!disableMethodTest2Hierarchy && isMethodTest && !callNode->getSymbolReference()->isUnresolved())
+             {
+             // If the profiled guard is a method test, get objectClass from the
+             // child of the lhs node.
+             TR::Node *loadNode = node->getFirstChild()->getFirstChild();
+             bool isGlobal = false;
+             TR::VPConstraint *loadConstraint = vp->getConstraint(loadNode, isGlobal);
+             objectClass = loadConstraint ? loadConstraint->getClass() : NULL;
+             isFixedClass = loadConstraint && loadConstraint->isFixedClass();
+
+             TR::Node *methodPtrNode = node->getSecondChild();
+             inlinedMethodClass = vp->comp()->fe()->getClassFromMethodBlock((TR_OpaqueMethodBlock*)methodPtrNode->getAddress());
+
+             TR_PersistentCHTable *chTable = vp->comp()->getPersistentInfo()->getPersistentCHTable();
+             TR_ResolvedMethod* resolvedMethod = callNode->getSymbolReference()->getSymbol()->getResolvedMethodSymbol()->getResolvedMethod();
```

Comment:
We should ask if the inlined method is overriden in object class' hierarchy, because the inlined method might be in different class hierarchy. You can get resolved method of the inlined method by calling `getInlinedResolvedMethod` with inlined site index (use `getCurrentInlinedSiteIndex`) from the profiled guard.